### DISCUSSION TITLE
new 32 layer model

### DISCUSSION THREAD

**Discussion by Jack000** [lucidrains/DALLE-pytorch#396] [Show and tell]
hey guys

following up on my previous post, I trained a 32 layer dalle-pytorch model. Here's the link https://github.com/Jack000/DALLE-pytorch/
I also put up the deepspeed checkpoint here: https://dall-3.com/models/dalle/

model settings:
- 32 layers, 12 heads
- text sequence length 64, image sequence length 256
- attn_types full
- shift_tokens and rotary_emb to True
- 16384 dict size for text tokens, 8192 dict size for image tokens
- gumbel f8 VQGAN (it has better embeddings with no "dead" tokens like the f16 versions)

I chose a size that would converge in a reasonable amount of time on 4x3090 (ie. months, not years) and this seemed to be about right.

For the learning rate I started with what I thought was a large value (1e-3) and halved it manually when the loss appeared to plateau, ending with 2e-5 after about a month. At this point I can't see any change in loss even with 0.999 ema on the loss curve, so I'm calling it done, but I suspect it's still under-trained.

For the dataset, I used a filtered version of Laion-400m and added some images I scraped myself, for a total of 60 million images. I filtered the data to include mostly photographic images (clip filtering is pretty noisy) and also to remove images with watermarks.

There are sample images in the github repo but here are some failure cases I noticed.

a green pentagonal clock. a green clock shaped like a pentagon

![clock](https://user-images.githubusercontent.com/2636509/144284214-f25547f5-e630-4d36-a34e-fdbc2a7aaa03.png)

seems to have trouble with shapes, but to be fair even OpenAI's DALL-E has trouble with this

barak obama appears at a press conference

![obama](https://user-images.githubusercontent.com/2636509/144284359-bd50a15c-0b1b-4257-b864-ab5b1247bf7f.png)

I think a larger transformer would help it memorize specific people/places

the birth of sentient ai

![birth](https://user-images.githubusercontent.com/2636509/144284446-78cd2efb-061d-4034-9dd4-490e8fb2754f.png)

If the prompt can be interpreted as an article title or news release it has a tendency to generate image-text. The data might need more aggressive filtering.

Overall I think you really need a larger transformer to get the generalization capabilities of OpenAI's DALL-E, but this model works pretty well for common object classes.

some notes on training:
- currently, changing learning rate on resume doesn't work with deepspeed (it'll take the previous learning rate from the checkpoint) I had to loop through the params manually
- not sure if it's just my system, but I ran into this issue: https://github.com/pytorch/pytorch/issues/13246 Swapping out the python native lists with pyarrow fixed it
- with deepspeed O1+amp I got NaN gradients and the scaling didn't seem to work (after the first NaN it never recovered), so I just used fp32-only training
- I tried attention_types full,sparse but it only increased throughput by 5% or so. Just went with full attention

anyways, I'll probably keep training this whenever my training rig is idle

**Comment by rom1504** [comment]
very cool stuff!!
we should put that model in a demo using https://github.com/rom1504/dalle-service or https://huggingface.co/spaces/rom1504/laion_car_subset/blob/main/app.py so people can try it out!

I'm curious to see more examples, for example the ones from https://openai.com/blog/dall-e/

**Comment by Jack000** [comment]
does huggingface spaces offer free gpu? That would be awesome since it has an actual gui, compared to colab

here are some more generations. Temp=0.9 top_p=0.85 - top 36 out of 256 images clip re-ranked. I didn't do the diffusion pass since it would take several days.

a cube made of porcupine. a cube with the texture of porcupine
![grid](https://user-images.githubusercontent.com/2636509/144324917-19a387ac-64a4-458d-9ccd-39ce94120859.png)

a collection of glasses is sitting on a table
![grid](https://user-images.githubusercontent.com/2636509/144325064-a0962a0d-12e9-4dbc-a895-a48f75fc578e.png)
seems like it's firmly decided on drinking glasses instead of eye glasses

a small red block sitting on a large green block
![grid](https://user-images.githubusercontent.com/2636509/144325218-fbc6e408-f623-49e5-a7be-e5d6bd9d4fcf.png)

a stack of 3 cubes. a red cube is on the top, sitting on a green cube. the green cube is in the middle, sitting on a blue cube. the blue cube is on the bottom
![grid](https://user-images.githubusercontent.com/2636509/144325589-71ea0247-020c-49aa-a521-10b6b79beb5c.png)
not quite, but it mostly gets the 3 colors

an extreme close-up view of a capybara sitting in a field
![grid](https://user-images.githubusercontent.com/2636509/144325745-12436cd7-4c42-49d9-a7a0-6efc19b39393.png)

a capybara made of voxels sitting in a field
![grid](https://user-images.githubusercontent.com/2636509/144325821-85506529-2d07-4ced-9fed-4879213ca692.png)
some of these do look more video-gamey

a cross sectional view of a walnut
![grid](https://user-images.githubusercontent.com/2636509/144325917-c2a7e38e-8eca-43f0-91c6-5b64c9d1df88.png)
not many cross sections. I wonder if it's a wording issue (maybe something like "inside of a walnut" or a more common phrasing would work better)

a macro photograph of brain coral
![grid](https://user-images.githubusercontent.com/2636509/144326130-d8178a8d-9fe4-484b-9ac2-54dcb20ee3ca.png)
this looks great, mostly because I have no idea what a brain coral is supposed to look like

a painting of a capybara sitting in a field at sunrise
![grid](https://user-images.githubusercontent.com/2636509/144326219-ad1c299d-1b2f-4d45-9ee4-2fb3b2f44e29.png)
I did filter out non-photographic images so it's not too surprising that they don't look very paintingish

a stained glass window with an image of a blue strawberry
![grid](https://user-images.githubusercontent.com/2636509/144326446-f3621c99-e585-4188-8b69-88034fbe68f7.png)

a storefront that has the word "openai" written on it. a store front with the word openai. openai storefront
![grid](https://user-images.githubusercontent.com/2636509/144326488-14e31b42-43e8-402b-b497-e33404b7028d.png)
it's trying, but there's not enough resolution to make legible text

a snail made of harp. a snail with the texture of harp
![grid](https://user-images.githubusercontent.com/2636509/144326621-29fe94ef-85ad-48cb-8928-638e18ddb6d5.png)
not great, maybe the diffusion pass would help with some of these

a photo of the food of china
![grid](https://user-images.githubusercontent.com/2636509/144326727-f58b6b93-e1ed-4e92-8b7f-d6857517c739.png)
a more generic prompt definitely works better. Some of these could be photo-realistic after the diffusion pass

a photo of alamo square, san francisco, from a street at night
![grid](https://user-images.githubusercontent.com/2636509/144326823-4c0d5c29-1476-483c-b4eb-df8f2a5bcdb4.png)

a photo of san francisco's golden gate bridge
![grid](https://user-images.githubusercontent.com/2636509/144326897-bc483454-9d50-4605-9769-e8539fa363b4.png)

a photo of a phone from the 1900s
![grid](https://user-images.githubusercontent.com/2636509/144326996-d9a14a84-de4d-4b3c-a7b2-8d16a7a5bb72.png)
hm.. maybe there are not enough historical photos in the dataset

a photo of a phone from the 2000s
![grid](https://user-images.githubusercontent.com/2636509/144327137-7822a89a-f2de-4370-9db6-d9579ece4333.png)
this seems closer to the mark

a photo of a phone from the distant future
![grid](https://user-images.githubusercontent.com/2636509/144327245-b3f87245-29ae-4e19-99c4-fe7d01e139da.png)

a professional high quality emoji of a lovestruck cup of boba
![grid](https://user-images.githubusercontent.com/2636509/144327395-5e7b3f28-4aea-4ec2-b501-3bac04c09e13.png)
not sure if it knows what a boba is, but it's definitely going for an illustrative style
